pkg/webhooks/podimagespec/podimagespec.go

```diff
@@ -0,0 +1,293 @@
+package podimagespec
+
+import (
+	"context"
+	"fmt"
+	"net/http"
+	"regexp"
+
+	"github.com/openshift/managed-cluster-validating-webhooks/pkg/k8sutil"
+	"github.com/openshift/managed-cluster-validating-webhooks/pkg/webhooks/utils"
+
+	imagestreamv1 "github.com/openshift/api/image/v1"
+	configv1 "github.com/openshift/api/imageregistry/v1"
+	registryv1 "github.com/openshift/api/imageregistry/v1"
+	operatorv1 "github.com/openshift/api/operator/v1"
+	admissionregv1 "k8s.io/api/admissionregistration/v1"
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/util/json"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	logf "sigs.k8s.io/controller-runtime/pkg/log"
+	admissionctl "sigs.k8s.io/controller-runtime/pkg/webhook/admission"
+)
+
+const (
+	WebhookName string = "podimagespec-mutation"
+	docString   string = `OpenShift debugging tools on Managed OpenShift clusters must be available even if internal image registry is removed.`
+)
+
+var (
+	timeout int32 = 2
+	scope         = admissionregv1.NamespacedScope
+	rules         = []admissionregv1.RuleWithOperations{
+		{
+			Operations: []admissionregv1.OperationType{
+				admissionregv1.Create,
+			},
+			Rule: admissionregv1.Rule{
+				APIGroups:   []string{""},
+				APIVersions: []string{"v1"},
+				Resources:   []string{"pods"},
+				Scope:       &scope,
+			},
+		},
+	}
+	log        = logf.Log.WithName(WebhookName)
+	imageRegex = regexp.MustCompile(`^(image-registry\.openshift-image-registry\.svc:5000\/)(?P<namespace>\S*)(/)(?P<image>\w*)(:)(?P<tag>\S*)`)
+)
+
+// PodImageSpecWebhook mutates an image spec in a pod
+type PodImageSpecWebhook struct {
+	s          *runtime.Scheme
+	kubeClient client.Client
+}
+
+// NewWebhook creates the new webhook
+func NewWebhook() *PodImageSpecWebhook {
+	scheme := runtime.NewScheme()
+	configv1.Install(scheme)
+	imagestreamv1.Install(scheme)
+	return &PodImageSpecWebhook{
+		s: scheme,
+	}
+}
+
+// Authorized implements Webhook interface
+func (s *PodImageSpecWebhook) Authorized(request admissionctl.Request) admissionctl.Response {
+	ret := s.authorized(request)
+	if err := ret.Complete(request); err != nil {
+		log.Error(err, "Failed to complete the request")
+		return admissionctl.Errored(http.StatusInternalServerError, err)
+	}
+	return ret
+}
+
+func (s *PodImageSpecWebhook) authorized(request admissionctl.Request) admissionctl.Response {
+	var err error
+	ctx := context.Background()
+
+	if s.kubeClient == nil {
+		s.kubeClient, err = k8sutil.KubeClient(s.s)
+		if err != nil {
+			return admissionctl.Errored(http.StatusBadRequest, err)
+		}
+	}
+
+	pod, err := s.renderPod(request)
+	if err != nil {
+		log.Error(err, "couldn't render a Pod from the incoming request")
+		return admissionctl.Errored(http.StatusBadRequest, err)
+	}
+
+	if !podContainsContainerRegexMatch(pod) {
+		return admissionctl.Allowed("Pod image spec is valid")
```

Comment:
Any time we return a response (besides bad requests), we need to ensure the `uid` is set, matching the request. There are two patterns in this repo:
* Doing it manually: https://github.com/openshift/managed-cluster-validating-webhooks/blob/8bda2d2b7709223945b04311e604c5a2b73e4f03/pkg/webhooks/service/service.go#L80-L81
* Using a helper: https://github.com/openshift/managed-cluster-validating-webhooks/blob/8bda2d2b7709223945b04311e604c5a2b73e4f03/pkg/webhooks/imagecontentpolicies/imagecontentpolicies.go#L55 / https://github.com/openshift/managed-cluster-validating-webhooks/blob/8bda2d2b7709223945b04311e604c5a2b73e4f03/pkg/webhooks/utils/utils.go#L116-L122